manifest-build-tools/application/tag_change_monitor.py

```diff
@@ -0,0 +1,181 @@
+        self.repo = repo
+        self.history_file = history_file
+        self.property_file = property_file
+        if credential:
+            user, password = os.environ[credential].split(':')
+            self.auth = (user, password)
+        else:
+            self.auth = ()
+
+        self.history = {}
+        self.new_history = {}
+        self.change = {}
+
+        if os.path.isfile(history_file):
+            with file(history_file, 'r') as f:
+                for tag_commit in f.readlines():
+                    commit = tag_commit.split()[0]
+                    tags = tag_commit.split()[1:]
+                    self.history[commit] = tags
+                f.close()
+        else:
+            with file(history_file, 'w') as f:
+                f.close()
+
+    def handle_tag_monitor(self):
+        """
+        Do tag monitor job
+        """
+        tag_list = self.get_tags()
+        self.parse_tag_info(tag_list)
+        if self.change:
+            self.write_property_file()
+        self.write_history_file()
+
+    def get_tags(self):
+        """
+        Get all tags of current repo.
+        """
+        list_tag_url = "/".join([self.api_url, "repos", self.repo, "tags"])
+        if self.auth:
+            resp = requests.get(list_tag_url, auth=self.auth)
+        else:
+            resp = requests.get(list_tag_url)
+        if resp.ok:
+            return resp.json()
+        else:
+            print "Get tags of repo {0} error.\n{1}".format(self.repo, resp.text)
+            sys.exit(1)
+
+    def parse_tag_info(self, tag_list):
+        """
+        Get tag change and generate new tag history record.
+        """
+        find_change = False
+        contains_in_new = True
+        for tag in tag_list:
+            tag_name = tag['name']
+            commit = tag['commit']['sha']
+
+            if (not self.history.has_key(commit)) or \
+                (tag_name not in self.history[commit]):
+                if not find_change:
+                    self.change[commit] = [tag_name]
+                    find_change = True
+                else:
+                    contains_in_new = False
+
+            if contains_in_new:
+                if self.new_history.has_key(commit):
+                    self.new_history[commit].append(tag_name)
+                else:
+                    self.new_history[commit] = [tag_name]
+            else:
+                contains_in_new = True
+
+    def write_history_file(self):
+        """
+        Write new_history to history_file
+        This can keep history file valid when deleting tags.
+        """
+        with file(self.history_file, 'w') as history_file_handler:
+            for commit, tags in self.new_history.iteritems():
+                tags.insert(0, commit)
+                tags.append('\n')
+                history_file_handler.write(" ".join(tags))
+
+    def write_property_file(self):
+        """
+        write property file for downstream job
```

Comment:
for those tags "newly" detected for this run,   write those tags' information into a property file ( being used by downstream Jenkins job )